### ISSUE TITLE
Installation is broken on Python 3.12

### ISSUE THREAD

**Issue by defreng** [pulumi/pulumi#14258]
### What happened?

currently, when trying to install the `pulumi` package in a Python 3.12 environment, it fails due to a build failure in the upstream `grpcio` dependency which is pinned to version `1.56.2`, which does not compile with Python 3.12.

On the grpcio side this was recently fixed and a new release is available:

https://github.com/grpc/grpc/issues/33063

### Example

-

### Output of `pulumi about`

-

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by justinvp** [comment]
Unfortunately, upgrading to grpcio 1.59 has caused diagnostic errors to be printed for Python dynamic providers: #14442. #14474 reverts #14259. We're blocked on supporting Python 3.12 until the upstream issue is addressed: https://github.com/grpc/grpc/issues/34853

**Comment by justinvp** [comment]
https://github.com/grpc/grpc/issues/34853 has been fixed. We're waiting on it to be released.

**Comment by marns93** [comment]
> [grpc/grpc#34853](https://github.com/grpc/grpc/issues/34853) has been fixed. We're waiting on it to be released.

@justinvp Any updates?

**Comment by Frassle** [comment]
> @justinvp Any updates?

grpc haven't made a release with this fix. Soon as they do we'll update.

**Comment by diogobaeder** [comment]
I just installed `grpcio` version `1.60.0` just fine with Python 3.12. Was it the right call to pin the dependency?

**Comment by justinvp** [comment]
We've worked around the upstream issue and Python 3.12 is now supported with v3.103.0 of the `pulumi` PyPi package. https://pypi.org/project/pulumi/

**Comment by diogobaeder** [comment]
Awesome job! Thanks, guys! :-)

**Comment by omidraha** [comment]
Same issue, after a long time with high  cpu usage `pulumi up` failed.

 `error: failed to discover plugin requirements: calling `python -m pip list -v --format json`: exit status 1`
```
pulumi version
v3.136.0
```

![Image](https://github.com/user-attachments/assets/2610a43a-3101-4f1f-9eb1-20729cadf2f5)

**Comment by justinvp** [comment]
@omidraha, would you mind opening a new issue?
